<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#E8_프로젝트---Movielens-영화-추천-실습" data-toc-modified-id="E8_프로젝트---Movielens-영화-추천-실습-1">E8_프로젝트 - Movielens 영화 추천 실습</a></span><ul class="toc-item"><li><span><a href="#0.-필요한-모듈-가져오기" data-toc-modified-id="0.-필요한-모듈-가져오기-1.1">0. 필요한 모듈 가져오기</a></span></li><li><span><a href="#1.-데이터-준비와-전처리" data-toc-modified-id="1.-데이터-준비와-전처리-1.2">1. 데이터 준비와 전처리</a></span><ul class="toc-item"><li><span><a href="#1-1.-Rating-file-준비-및-전처리" data-toc-modified-id="1-1.-Rating-file-준비-및-전처리-1.2.1">1-1. Rating file 준비 및 전처리</a></span></li><li><span><a href="#1-2.-Movie-file-준비-및-전처리" data-toc-modified-id="1-2.-Movie-file-준비-및-전처리-1.2.2">1-2. Movie file 준비 및 전처리</a></span></li></ul></li><li><span><a href="#2.-데이터-분석" data-toc-modified-id="2.-데이터-분석-1.3">2. 데이터 분석</a></span></li><li><span><a href="#3.-모델-검증을-위한-사용자-초기-정보-세팅" data-toc-modified-id="3.-모델-검증을-위한-사용자-초기-정보-세팅-1.4">3. 모델 검증을 위한 사용자 초기 정보 세팅</a></span><ul class="toc-item"><li><span><a href="#3-1.-내가-좋아하는-영화-5가지-추가" data-toc-modified-id="3-1.-내가-좋아하는-영화-5가지-추가-1.4.1">3-1. 내가 좋아하는 영화 5가지 추가</a></span></li><li><span><a href="#3-2.-모델에-활용하기-위한-전처리" data-toc-modified-id="3-2.-모델에-활용하기-위한-전처리-1.4.2">3-2. 모델에 활용하기 위한 전처리</a></span></li><li><span><a href="#3-3.-indexing을-통해-데이터-컬럼-내-값을-바꾸기" data-toc-modified-id="3-3.-indexing을-통해-데이터-컬럼-내-값을-바꾸기-1.4.3">3-3. indexing을 통해 데이터 컬럼 내 값을 바꾸기</a></span></li></ul></li><li><span><a href="#4.-CSR-matrix를-직접-만들어-봅시다." data-toc-modified-id="4.-CSR-matrix를-직접-만들어-봅시다.-1.5">4. CSR matrix를 직접 만들어 봅시다.</a></span></li><li><span><a href="#5.-als_model-=-AlternatingLeastSquares-모델을-직접-구성하여-훈련시켜-봅시다." data-toc-modified-id="5.-als_model-=-AlternatingLeastSquares-모델을-직접-구성하여-훈련시켜-봅시다.-1.6">5. als_model = AlternatingLeastSquares 모델을 직접 구성하여 훈련시켜 봅시다.</a></span><ul class="toc-item"><li><span><a href="#5-1.-모델-선언" data-toc-modified-id="5-1.-모델-선언-1.6.1">5-1. 모델 선언</a></span></li><li><span><a href="#5-2.-모델-학습" data-toc-modified-id="5-2.-모델-학습-1.6.2">5-2. 모델 학습</a></span></li></ul></li><li><span><a href="#6.-내가-선호하는-5가지-영화-중-하나와-그-외의-영화-하나를-골라-훈련된-모델이-예측한-나의-선호도를-파악해-보세요." data-toc-modified-id="6.-내가-선호하는-5가지-영화-중-하나와-그-외의-영화-하나를-골라-훈련된-모델이-예측한-나의-선호도를-파악해-보세요.-1.7">6. 내가 선호하는 5가지 영화 중 하나와 그 외의 영화 하나를 골라 훈련된 모델이 예측한 나의 선호도를 파악해 보세요.</a></span></li><li><span><a href="#7.-내가-좋아하는-영화와-비슷한-영화를-추천받아-봅시다." data-toc-modified-id="7.-내가-좋아하는-영화와-비슷한-영화를-추천받아-봅시다.-1.8">7. 내가 좋아하는 영화와 비슷한 영화를 추천받아 봅시다.</a></span></li><li><span><a href="#8.-내가-가장-좋아할-만한-영화들을-추천받아-봅시다." data-toc-modified-id="8.-내가-가장-좋아할-만한-영화들을-추천받아-봅시다.-1.9">8. 내가 가장 좋아할 만한 영화들을 추천받아 봅시다.</a></span></li><li><span><a href="#회고" data-toc-modified-id="회고-1.10">회고</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#1.-pd.read_csv()에서-에러가-발생함" data-toc-modified-id="1.-pd.read_csv()에서-에러가-발생함-1.10.0.1">1. pd.read_csv()에서 에러가 발생함</a></span></li><li><span><a href="#2..groupby('movie_id')['user_id'].count()" data-toc-modified-id="2..groupby('movie_id')['user_id'].count()-1.10.0.2">2.<code>.groupby('movie_id')['user_id'].count()</code></a></span></li><li><span><a href="#3.-판다스-데이터프레임에서-데이터-전처리하기" data-toc-modified-id="3.-판다스-데이터프레임에서-데이터-전처리하기-1.10.0.3">3. 판다스 데이터프레임에서 데이터 전처리하기</a></span></li><li><span><a href="#4.-AlternatingLeastSquares-모델을-사용하면서" data-toc-modified-id="4.-AlternatingLeastSquares-모델을-사용하면서-1.10.0.4">4. AlternatingLeastSquares 모델을 사용하면서</a></span></li></ul></li></ul></li></ul></li></ul></div>

# E8_프로젝트 - Movielens 영화 추천 실습

**루브릭**

아래의 기준을 바탕으로 프로젝트를 평가합니다.

| 평가문항 |상세기준 |  
|:----------|:----------|
| 1. CSR matrix가 정상적으로 만들어졌다. |사용자와 아이템 개수를 바탕으로 정확한 사이즈로 만들었다. |  
| 2. MF 모델이 정상적으로 훈련되어 그럴듯한 추천이 이루어졌다. |사용자와 아이템 벡터 내적수치가 의미있게 형성되었다. |  
| 3. 비슷한 영화 찾기와 유저에게 추천하기의 과정이 정상적으로 진행되었다. |MF모델이 예측한 유저 선호도 및 아이템간 유사도, 기여도가 의미있게 측정되었다. |

이전 스텝에서 배운 MF 모델 학습 방법을 토대로, 내가 좋아할 만한 영화 추천 시스템을 제작해 보겠습니다.  
이번에 활용할 데이터셋은 **추천시스템의 MNIST**라고 부를만한 **Movielens 데이터**입니다.

* 유저가 영화에 대해 평점을 매긴 데이터가 데이터 크기 별로 있습니다. MovieLens 1M Dataset 사용을 권장합니다.
* 별점 데이터는 대표적인 explicit 데이터입니다. 하지만 implicit 데이터로 간주하고 테스트해볼 수 있습니다.
* 별점을 시청횟수로 해석해서 생각하겠습니다.
* 또한 유저가 3점 미만으로 준 데이터는 선호하지 않는다고 가정하고 제외하겠습니다.

## 0. 필요한 모듈 가져오기

In [1]:
import pandas as pd
import numpy as np
import os

## 1. 데이터 준비와 전처리
Movielens 데이터는 rating.dat 안에 이미 인덱싱까지 완료된 사용자-영화-평점 데이터가 깔끔하게 정리되어 있습니다.

### 1-1. Rating file 준비 및 전처리
**RATINGS FILE DESCRIPTION**  
All ratings are contained in the file "ratings.dat" and are in the following format:

<span style="color:blue">**UserID::MovieID::Rating::Timestamp**</span>


- UserIDs range between 1 and 6040 
- MovieIDs range between 1 and 3952
- Ratings are made on a 5-star scale (whole-star ratings only)
- Timestamp is represented in seconds since the epoch as returned by time(2)
- Each user has at least 20 ratings

In [2]:
rating_file_path=os.getenv('HOME') + '/aiffel/recommendata_iu/data/ml-1m/ratings.dat'
ratings_cols = ['user_id', 'movie_id', 'rating', 'timestamp']
ratings = pd.read_csv(rating_file_path, sep='::', names=ratings_cols, engine='python')
orginal_data_size = len(ratings)
ratings.head()

user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291

In [3]:
# 3점 이상만 남깁니다.
ratings = ratings[ratings['rating']>=3]
filtered_data_size = len(ratings)

print(f'orginal_data_size: {orginal_data_size}, filtered_data_size: {filtered_data_size}')
print(f'Ratio of Remaining Data is {filtered_data_size / orginal_data_size:.2%}')

orginal_data_size: 1000209, filtered_data_size: 836478
Ratio of Remaining Data is 83.63%


In [4]:
# rating 컬럼의 이름을 count로 바꿉니다.
ratings.rename(columns={'rating':'count'}, inplace=True)
ratings.head()

user_id  movie_id  count  timestamp
0        1      1193      5  978300760
1        1       661      3  978302109
2        1       914      3  978301968
3        1      3408      4  978300275
4        1      2355      5  978824291

### 1-2. Movie file 준비 및 전처리
**MOVIES FILE DESCRIPTION**  
Movie information is in the file "movies.dat" and is in the following
format:  

<span style='color:blue'>**MovieID::Title::Genres**</span>


- Titles are identical to titles provided by the IMDB (including
year of release)
- Genres are pipe-separated and are selected from the following genres:

	* Action
	* Adventure
	* Animation
	* Children's
	* Comedy
	* Crime
	* Documentary
	* Drama
	* Fantasy
	* Film-Noir
	* Horror
	* Musical
	* Mystery
	* Romance
	* Sci-Fi
	* Thriller
	* War
	* Western

- Some MovieIDs do not correspond to a movie due to accidental duplicate
entries and/or test entries
- Movies are mostly entered by hand, so errors and inconsistencies may exist

In [5]:
# 영화 제목을 보기 위해 메타 데이터를 읽어옵니다.
movie_file_path=os.getenv('HOME') + '/aiffel/recommendata_iu/data/ml-1m/movies.dat'
movies_cols = ['movie_id', 'title', 'genre'] 
movies = pd.read_csv(movie_file_path, sep='::', names=movies_cols, engine='python', encoding='ISO-8859-1')
movies.head()

movie_id                               title                         genre
0         1                    Toy Story (1995)   Animation|Children's|Comedy
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy
2         3             Grumpier Old Men (1995)                Comedy|Romance
3         4            Waiting to Exhale (1995)                  Comedy|Drama
4         5  Father of the Bride Part II (1995)                        Comedy

<span style='color:blue'>**UTF-8과 ISO-8859-1의 차이점**</sppan>  
* **UTF-8**은 모든 유니 코드 문자를 나타낼 수있는 **멀티 바이트 인코딩**입니다. 
* **ISO 8859-1**은 처음 256 개의 유니 코드 문자를 나타낼 수있는 **1 바이트 인코딩**입니다. 
* 둘 다 ASCII를 정확히 같은 방식으로 인코딩합니다.

In [6]:
# 검색을 쉽게하기 위해 영화제목과 장르의 문자열을 소문자로 바꾸기
movies['title'] = movies['title'].str.lower() 
movies['genre'] = movies['genre'].str.lower() 
movies.head()

movie_id                               title                         genre
0         1                    toy story (1995)   animation|children's|comedy
1         2                      jumanji (1995)  adventure|children's|fantasy
2         3             grumpier old men (1995)                comedy|romance
3         4            waiting to exhale (1995)                  comedy|drama
4         5  father of the bride part ii (1995)                        comedy

## 2. 데이터 분석
* ratings에 있는 유니크한 영화 개수
* ratings에 있는 유니크한 사용자 수
* 가장 인기 있는 영화 10개(인기순)

In [7]:
# ratings에 있는 유니크한 영화 개수
ratings['movie_id'].nunique()

3628

In [8]:
# ratings에 있는 유니크한 사용자 수
ratings['user_id'].nunique()

6039

In [9]:
# 가장 인기 있는 영화 10개(인기순)
movie_count = ratings.groupby('movie_id')['user_id'].count()
movie_count.sort_values(ascending=False).head(10)

movie_id
2858    3211
260     2910
1196    2885
1210    2716
2028    2561
589     2509
593     2498
1198    2473
1270    2460
2571    2434
Name: user_id, dtype: int64

**`.groupby('movie_id')['user_id'].count()`**  
* 데이터를 'movie_id' 기준으로 그룹화하여 개수를 집계하는데 'user_id' 컬럼만 수행한다. 

**`.sort_values(ascending=False)`**
* 데이터를 정렬하는데 값을 기준으로 오름차순이 아니라 내림차순으로 수행한다.

In [10]:
# 첫 번째 유저가 어떤 영화를 보는지 확인
condition = (ratings['user_id'] == ratings.loc[0, 'user_id'])
ratings.loc[condition]

user_id  movie_id  count  timestamp
0         1      1193      5  978300760
1         1       661      3  978302109
2         1       914      3  978301968
3         1      3408      4  978300275
4         1      2355      5  978824291
5         1      1197      3  978302268
6         1      1287      5  978302039
7         1      2804      5  978300719
8         1       594      4  978302268
9         1       919      4  978301368
10        1       595      5  978824268
11        1       938      4  978301752
12        1      2398      4  978302281
13        1      2918      4  978302124
14        1      1035      5  978301753
15        1      2791      4  978302188
16        1      2687      3  978824268
17        1      2018      4  978301777
18        1      3105      5  978301713
19        1      2797      4  978302039
20        1      2321      3  978302205
21        1       720      3  978300760
22        1      1270      5  978300055
23        1       527      5  978824195
24        1      2340      3  978300103
25        1        48      5  978824351
26        1      1097      4  978301953
27        1      1721      4  978300055
28        1      1545      4  978824139
29        1       745      3  978824268
30        1      2294      4  978824291
31        1      3186      4  978300019
32        1      1566      4  978824330
33        1       588      4  978824268
34        1      1907      4  978824330
35        1       783      4  978824291
36        1      1836      5  978300172
37        1      1022      5  978300055
38        1      2762      4  978302091
39        1       150      5  978301777
40        1         1      5  978824268
41        1      1961      5  978301590
42        1      1962      4  978301753
43        1      2692      4  978301570
44        1       260      4  978300760
45        1      1028      5  978301777
46        1      1029      5  978302205
47        1      1207      4  978300719
48        1      2028      5  978301619
49        1       531      4  978302149
50        1      3114      4  978302174
51        1       608      4  978301398
52        1      1246      4  978302091

In [11]:
# 유저별로 몇 편의 영화를 보고 있는지에 대한 통계
user_count = ratings.groupby('user_id')['movie_id'].count()
user_count.describe()

count    6039.000000
mean      138.512668
std       156.241599
min         1.000000
25%        38.000000
50%        81.000000
75%       177.000000
max      1968.000000
Name: movie_id, dtype: float64

6039명의 유저가 있으며, 유저들은 데이터 집계기간 동안에 81편(중간값)의 영화를 보았다.

In [12]:
# 유저별 count횟수 중앙값에 대한 통계
user_median = ratings.groupby('user_id')['count'].median()
user_median.describe()

count    6039.000000
mean        4.055970
std         0.432143
min         3.000000
25%         4.000000
50%         4.000000
75%         4.000000
max         5.000000
Name: count, dtype: float64

6039명의 유저가 있으며, 유저들은 대략 4회 정도 영화를 보았다.

## 3. 모델 검증을 위한 사용자 초기 정보 세팅

### 3-1. 내가 좋아하는 영화 5가지 추가

In [13]:
# 본인이 좋아하는 영화 5가지 추가, 이름은 꼭 데이터셋에 있는 것과 동일하게 맞춰주세요. 
my_favorite = ['apollo 13 (1995)', 'waterworld (1995)', 'before sunrise (1995)', 'stargate (1994)', 'batman (1989)']

# 'yongbok'이라는 user_id가 위 영화 5개를 4회씩 보았다고 가정하겠습니다.
my_playlist = pd.DataFrame({'user_id': ['yongbok']*5, 'movie_id': my_favorite, 'count':[4]*5})

# my_playlist를 기존 ratings 데이터에 추가
if not ratings.isin({'user_id': ['yongbok']})['user_id'].any():    # user_id에 'yongbok' 또는 6041 이라는 데이터가 없다면
    ratings = ratings.append(my_playlist)                     # 위에 임의로 만든 my_favorite 데이터를 추가해 줍니다. 

ratings.tail(10)       # 잘 추가되었는지 확인해 봅시다.

user_id               movie_id  count    timestamp
1000203     6040                   1090      3  956715518.0
1000205     6040                   1094      5  956704887.0
1000206     6040                    562      5  956704746.0
1000207     6040                   1096      4  956715648.0
1000208     6040                   1097      4  956715569.0
0        yongbok       apollo 13 (1995)      4          NaN
1        yongbok      waterworld (1995)      4          NaN
2        yongbok  before sunrise (1995)      4          NaN
3        yongbok        stargate (1994)      4          NaN
4        yongbok          batman (1989)      4          NaN

### 3-2. 모델에 활용하기 위한 전처리 

In [14]:
# 고유한 값 찾기
movie_unique = movies['title'].unique()
user_unique = ratings['user_id'].unique()

# 고유한 값을 indexing 하기 
movie_to_idx = {v:k+1 for k, v in enumerate(movie_unique)}    # k: key, v: value
user_to_idx = {v:k+1 for k, v in enumerate(user_unique)}

# indexing으로 고유한 값 찾기
idx_to_movie = {k+1:v for k, v in enumerate(movie_unique)}
idx_to_user = {k+1:v for k, v in enumerate(user_unique)}

In [15]:
# 인덱싱이 잘 되었는지 확인
my_idx = user_to_idx['yongbok']
print(f'user_id of yongbok is {my_idx}')
print(f'user {my_idx} is {idx_to_user[my_idx]}')
print()
mv_idx = movie_to_idx['apollo 13 (1995)']
print(f'movie_id of apollo 13 (1995) is {mv_idx}')
print(f'movie {mv_idx} is {idx_to_movie[mv_idx]}')

user_id of yongbok is 6040
user 6040 is yongbok

movie_id of apollo 13 (1995) is 149
movie 149 is apollo 13 (1995)


### 3-3. indexing을 통해 데이터 컬럼 내 값을 바꾸기

In [16]:
# user_to_idx.get을 통해 user_id 컬럼의 모든 값을 인덱싱한 Series를 구해 봅시다. 
# 혹시 정상적으로 인덱싱되지 않은 row가 있다면 인덱스가 NaN이 될 테니 dropna()로 제거합니다. 
temp_user_data = ratings['user_id'].map(user_to_idx.get).dropna()

if len(temp_user_data) == len(ratings):   # 모든 row가 정상적으로 인덱싱되었다면
    print('user_id column indexing OK!!')
    ratings['user_id'] = temp_user_data   # data['user_id']을 인덱싱된 Series로 교체해 줍니다. 
else:
    print('user_id column indexing Fail!!')

    
# movie_to_idx을 통해 artist 컬럼도 동일한 방식으로 인덱싱해 줍니다. 
temp_movie_data = ratings['movie_id'].map(movie_to_idx.get).dropna()

if len(temp_movie_data) == len(ratings):
    print('movie_id column indexing OK!!')
    ratings['movie_id'] = temp_movie_data
else:
    print('movie_id column indexing Fail!!')

ratings.tail(10)

user_id column indexing OK!!
movie_id column indexing Fail!!


user_id               movie_id  count    timestamp
1000203     6039                   1090      3  956715518.0
1000205     6039                   1094      5  956704887.0
1000206     6039                    562      5  956704746.0
1000207     6039                   1096      4  956715648.0
1000208     6039                   1097      4  956715569.0
0           6040       apollo 13 (1995)      4          NaN
1           6040      waterworld (1995)      4          NaN
2           6040  before sunrise (1995)      4          NaN
3           6040        stargate (1994)      4          NaN
4           6040          batman (1989)      4          NaN

In [17]:
# movie_id 컬럼에 영화 제목으로 남아 있는 것을 인덱스로 변환
ratings.loc[(ratings['user_id']==6040), 'movie_id'] = ratings['movie_id'].map(movie_to_idx.get).dropna()
ratings.tail(10)

user_id movie_id  count    timestamp
1000203     6039     1090      3  956715518.0
1000205     6039     1094      5  956704887.0
1000206     6039      562      5  956704746.0
1000207     6039     1096      4  956715648.0
1000208     6039     1097      4  956715569.0
0           6040    149.0      4          NaN
1           6040    207.0      4          NaN
2           6040    214.0      4          NaN
3           6040    314.0      4          NaN
4           6040    589.0      4          NaN

In [18]:
# timestamp의 NaN은 평균값으로 대체
ratings.loc[(ratings['user_id']==6040), 'timestamp'] = int(ratings['timestamp'].mean())
ratings.tail(10) 

user_id movie_id  count    timestamp
1000203     6039     1090      3  956715518.0
1000205     6039     1094      5  956704887.0
1000206     6039      562      5  956704746.0
1000207     6039     1096      4  956715648.0
1000208     6039     1097      4  956715569.0
0           6040    149.0      4  972162834.0
1           6040    207.0      4  972162834.0
2           6040    214.0      4  972162834.0
3           6040    314.0      4  972162834.0
4           6040    589.0      4  972162834.0

## 4. CSR matrix를 직접 만들어 봅시다.

In [19]:
from scipy.sparse import csr_matrix

num_user = ratings['user_id'].nunique()
num_movie = ratings['movie_id'].nunique()

num_user, num_movie

(6040, 3628)

In [20]:
# csr_data = csr_matrix((ratings['count'], (ratings['user_id'], ratings['movie_id'])), shape= (num_user, num_movie))
csr_data = csr_matrix((ratings['count'], (ratings['user_id'], ratings['movie_id'])), shape=None)
csr_data.shape

(6041, 3953)

## 5. als_model = AlternatingLeastSquares 모델을 직접 구성하여 훈련시켜 봅시다.

In [21]:
from implicit.als import AlternatingLeastSquares
import os
import numpy as np

# implicit 라이브러리에서 권장하고 있는 부분입니다. 학습 내용과는 무관합니다.
os.environ['OPENBLAS_NUM_THREADS']='1'
os.environ['KMP_DUPLICATE_LIB_OK']='True'
os.environ['MKL_NUM_THREADS']='1'

### 5-1. 모델 선언

In [22]:
# Implicit AlternatingLeastSquares 모델의 선언
als_model = AlternatingLeastSquares(factors=100, 
                                    regularization=0.01, 
                                    use_gpu=False, 
                                    iterations=15, 
                                    dtype=np.float32)

**AlternatingLeastSquares 클래스**의 **__init__ 파라미터**   
**1. factors :** 유저와 아이템의 벡터를 몇 차원으로 할 것인지   
**2. regularization :** 과적합을 방지하기 위해 정규화 값을 얼마나 사용할 것인지   
**3. use_gpu :** GPU를 사용할 것인지   
**4. iterations :** epochs와 같은 의미입니다. 데이터를 몇 번 반복해서 학습할 것인지  

1,4를 늘릴수록 학습데이터를 잘 학습하게 되지만 과적합의 우려가 있으니 좋은 값을 찾아야 함

In [23]:
# als 모델은 input으로 (item X user 꼴의 matrix를 받기 때문에 Transpose해줍니다.)
csr_data_transpose = csr_data.T
csr_data_transpose

<3953x6041 sparse matrix of type '<class 'numpy.int64'>'
	with 836483 stored elements in Compressed Sparse Column format>

### 5-2. 모델 학습

In [24]:
# 모델 학습
als_model.fit(csr_data_transpose)

  0%|          | 0/15 [00:00<?, ?it/s]

## 6. 내가 선호하는 5가지 영화 중 하나와 그 외의 영화 하나를 골라 훈련된 모델이 예측한 나의 선호도를 파악해 보세요.

In [25]:
# 모델이 저의 벡터와 apollo 13 (1995)의 벡터를 어떻게 만들고 있는지
yongbok, apollo13 = user_to_idx['yongbok'], movie_to_idx['apollo 13 (1995)']

yongbok_vector = als_model.user_factors[yongbok]

apollo13_vector = als_model.item_factors[apollo13]

* **`.user_factors[ ]`** : 훈련 세트의 각 **사용자**에 대한 잠재 요인 배열

* **`.item_factors[ ]`** : 훈련 세트의 각 **항목**에 대한 잠재 요인 배열

In [26]:
# yongbok의 벡터
yongbok_vector

array([-0.2464316 ,  0.12403583,  0.4124863 ,  0.29052123,  0.39670518,
       -0.06623028, -0.22203402,  0.03080336,  0.5847165 ,  0.11868675,
       -0.17690079,  0.1803783 , -0.34483853,  0.13232294,  0.10560482,
        0.277365  , -0.87117285, -0.09259569, -0.23296317, -0.05552815,
        0.34846047, -0.01694998,  0.01364768,  0.21485047,  0.02636858,
       -0.01141187,  0.2361918 ,  0.23940812,  0.26837754,  0.18735859,
        0.23078404, -0.1509557 ,  0.18200544,  0.11374459, -0.31014055,
        0.2343555 , -0.3455568 , -0.15373392,  0.22069855, -0.40859196,
       -0.02503973, -0.61159503, -0.14619476, -0.09692829, -0.27959946,
        0.22195731,  0.34412003,  0.56660694, -0.21983035,  0.05180404,
       -0.19382717,  0.5166889 , -0.37051508,  0.19855334,  0.2566444 ,
        0.27408165,  0.21549642,  0.3303353 , -0.11395265,  0.0678273 ,
        0.06828269, -0.14804639,  0.11368599,  0.06737231, -0.1269726 ,
       -0.04378165,  0.1317719 , -0.5636908 ,  0.14958267, -0.06

In [27]:
# apollo13의 벡터
apollo13_vector

array([-0.00333993,  0.00686694,  0.00564461,  0.00440952,  0.00444247,
        0.00479493, -0.0044789 , -0.00661483,  0.01071008, -0.00307123,
        0.01747457,  0.00632362,  0.00481173,  0.00692363,  0.00094019,
       -0.00087817, -0.00185108,  0.01424395, -0.00408331,  0.00498029,
        0.00732388,  0.00530185,  0.00918176,  0.00449032,  0.00944753,
        0.00649573, -0.00103741,  0.00392232,  0.01026496,  0.0181318 ,
        0.00841316,  0.00156357, -0.00202455,  0.0016412 , -0.00123054,
        0.00809515,  0.00757567,  0.0036723 ,  0.01277316,  0.01013258,
        0.01093805, -0.00599599,  0.01094101, -0.0036562 ,  0.01004784,
        0.00457397,  0.00318543, -0.00083287,  0.00669824,  0.00557218,
        0.01208937,  0.01255197, -0.00342597,  0.00511944,  0.00103467,
        0.00581756, -0.00165856,  0.00563676,  0.01182675,  0.01048841,
       -0.00197241,  0.00108647, -0.0061211 ,  0.01239236,  0.00478824,
       -0.00288295,  0.00118624, -0.00565647, -0.0090806 ,  0.00

In [28]:
# yongbok과 apollo13 백터를 내적
np.dot(yongbok_vector, apollo13_vector)

0.04067733

In [29]:
# 모델이 yongbok의 'dragonheart (1996)' 대한 선호도를 어떻게 예측
movie = movie_to_idx['dragonheart (1996)']
movie_vector = als_model.item_factors[movie]
np.dot(yongbok_vector, movie_vector)

0.095290646

## 7. 내가 좋아하는 영화와 비슷한 영화를 추천받아 봅시다.

In [30]:
# 'stargate (1994)'와 비슷한 영화 찾기
favorite_movie = 'stargate (1994)'
movie_id = movie_to_idx[favorite_movie]
similar_movie = als_model.similar_items(movie_id, N=15)       # (movie_id, 유사도) Tuple 로 반환
similar_movie

[(314, 0.9999999),
 (43, 0.5673457),
 (82, 0.55746),
 (301, 0.54960495),
 (3079, 0.53974694),
 (1695, 0.53688556),
 (2609, 0.53226334),
 (2493, 0.53082055),
 (1177, 0.52784985),
 (2310, 0.52712303),
 (1914, 0.5145214),
 (2390, 0.513638),
 (421, 0.5108783),
 (1650, 0.5035645),
 (531, 0.50301677)]

**`.similar_items`** : 유사한 항목 목록을 계산

In [31]:
# movie_to_idx 를 뒤집어, index로부터 movie을 얻는 dict를 생성합니다. 
idx_to_movie = {v:k for k,v in movie_to_idx.items()}
[idx_to_movie[i[0]] for i in similar_movie]

['stargate (1994)',
 'restoration (1995)',
 "antonia's line (antonia) (1995)",
 'quick and the dead, the (1995)',
 'green mile, the (1999)',
 'senseless (1998)',
 'buena vista social club (1999)',
 'true crime (1999)',
 "one flew over the cuckoo's nest (1975)",
 'police academy (1984)',
 'halloween (1978)',
 'armed and dangerous (1986)',
 'blue chips (1994)',
 'bent (1997)',
 'shadowlands (1993)']

In [32]:
# 비슷한 영화를 찾는 함수
def get_similar_movie(movie_name: str):
    movie_id = movie_to_idx[movie_name]
    similar_movie = als_model.similar_items(movie_id)
    similar_movie = [idx_to_movie[i[0]] for i in similar_movie]
    return similar_movie

In [33]:
# jumanji (1995)과 비슷한 영화 찾기 확인
get_similar_movie('jumanji (1995)')

['jumanji (1995)',
 'jennifer 8 (1992)',
 'indian in the cupboard, the (1995)',
 "billy's holiday (1995)",
 'run of the country, the (1995)',
 'fallen (1998)',
 'always tell your wife (1923)',
 'hello mary lou: prom night ii (1987)',
 'shallow grave (1994)',
 'buffalo 66 (1998)']

## 8. 내가 가장 좋아할 만한 영화들을 추천받아 봅시다.

In [34]:
user = user_to_idx['yongbok']

# recommend에서는 user*item CSR Matrix를 받습니다.
movie_recommended = als_model.recommend(user, csr_data, N=20, filter_already_liked_items=True)
movie_recommended

[(2571, 0.26774198),
 (480, 0.25646093),
 (1580, 0.21149796),
 (2916, 0.19893314),
 (2762, 0.17915478),
 (1240, 0.16465196),
 (110, 0.1641268),
 (1214, 0.14009666),
 (457, 0.13899533),
 (1249, 0.13639444),
 (593, 0.11139741),
 (1035, 0.10898399),
 (524, 0.107426085),
 (1544, 0.10741632),
 (529, 0.10256505),
 (1610, 0.1021902),
 (266, 0.10142901),
 (2791, 0.10125345),
 (1914, 0.10048701),
 (2291, 0.098945305)]

**`.recommend`** : 사용자에 대한 N 개의 권장 사항을 계산하고 항목 ID, 점수 목록을 반환

In [35]:
# 추천한 영화의 index로부터 영화의 제목 확인
[idx_to_movie[i[0]] for i in movie_recommended]

['mommie dearest (1981)',
 'king of the hill (1993)',
 'kicked in the head (1997)',
 'on any sunday (1971)',
 'cabaret balkan (bure baruta) (1998)',
 'stand by me (1986)',
 'taxi driver (1976)',
 'stalker (1979)',
 'getting even with dad (1994)',
 'pump up the volume (1990)',
 'pinocchio (1940)',
 'long kiss goodnight, the (1996)',
 "schindler's list (1993)",
 'contact (1997)',
 'serial mom (1994)',
 'phantoms (1998)',
 'little odessa (1994)',
 'stop making sense (1984)',
 'halloween (1978)',
 'waking ned devine (1998)']

In [36]:
# 'king of the hill (1993)'영화를 추천한 이유는?
item_id = movie_to_idx['king of the hill (1993)']
explain = als_model.explain(user, csr_data, itemid=item_id)
explain

(0.25450010039692705,
 [(589, 0.2404832948380863),
  (314, 0.013655178831279934),
  (149, 0.0042148152712441055),
  (207, -0.0018365975817212412),
  (214, -0.0020165909619620503)],
 (array([[ 0.6095791 ,  0.0883569 ,  0.12962498, ...,  0.06861631,
           0.0905824 ,  0.11854385],
         [ 0.05386052,  0.64432866,  0.0366195 , ...,  0.05562773,
           0.10368962,  0.05445039],
         [ 0.07901668,  0.03504826,  0.62295248, ...,  0.08475724,
           0.08579749,  0.13297416],
         ...,
         [ 0.04182707,  0.04190527,  0.06373118, ...,  0.54810669,
           0.00668587, -0.0528581 ],
         [ 0.05521714,  0.07481377,  0.06898656, ...,  0.05791606,
           0.55571269,  0.02210833],
         [ 0.07226186,  0.04555812,  0.10019678, ...,  0.02252277,
           0.08991489,  0.56832947]]),
  False))

**`.explain()`** : 사용자가 항목을 좋아하는 이유에 대한 설명을 제공

In [37]:
# 내가 기록으로 남긴 데이터 중 이번 'king of the hill (1993)'추천에 기여한 정도를 확인 
[(idx_to_movie[i[0]], i[1]) for i in explain[1]]

[('batman (1989)', 0.2404832948380863),
 ('stargate (1994)', 0.013655178831279934),
 ('apollo 13 (1995)', 0.0042148152712441055),
 ('waterworld (1995)', -0.0018365975817212412),
 ('before sunrise (1995)', -0.0020165909619620503)]

'king of the hill (1993)'을 추천한 이유는 내가 기록에 남긴 'batman (1989)' 영화 때문이구나...  
'king of the hill (1993)'을 한번 검색해서 한번 봐야겠다.

## 회고
#### 1. pd.read_csv()에서 에러가 발생함 
* 에러 코드 : <span style='color:red'>UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe9 in position 3114: invalid continuation byte</span>   
* `encoding='utf-8'` 을 사용해도 문제가 해결되지 않음  
* `encoding='ISO-8859-1'`를 이용하여 문제 해결  

#### 2.` .groupby('movie_id')['user_id'].count()`
* 데이터를 'movie_id' 기준으로 그룹화하여 개수를 집계(count())하는데 'user_id' 컬럼만 수행  

#### 3. 판다스 데이터프레임에서 데이터 전처리하기  
* 특정 조건에 맞는 행에 대해서 변경하고자 하는 컬럼 값을 하나의 값으로 변경하기  
* 특정 조건에 맞는 행에 대해서 변경하고자 하는 컬럼 값을 함수 적용한 값으로 변경하기  

#### 4. AlternatingLeastSquares 모델을 사용하면서 
* 어떻게 계산되는지 잘 모르겠으나 모델이 똑똑하다고 느꼈다.
* matrix factorization 에 대한 내용을 이해하는데 같은 조원의 설명이 많은 도움이 되었다.
* `csr_matrix((ratings['count'], (ratings['user_id'], ratings['movie_id'])), shape= (num_user, num_movie))`에서 
`TypeError: len() of unsized object`에러가 발생하는 이유는 찾지 못했음.(`shape=None`으로 변경함)
